File: app/src/main/java/com/example/data/SkinProject.kt
package com.example.data

import androidx.room.Entity
import androidx.room.PrimaryKey

@Entity(tableName = "skin_projects")
data class SkinProject(
    @PrimaryKey(autoGenerate = true) val id: Int = 0,
    val name: String,
    val format: String = "64x64", // "64x32", "64x64", "128x128"
    val modelType: String = "STEVE", // "STEVE", "ALEX"
    val textureData: ByteArray, // PNG image bytes
    val folder: String = "My Skins",
    val isFavorite: Boolean = false,
    val lastUpdated: Long = System.currentTimeMillis()
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is SkinProject) return false

        if (id != other.id) return false
        if (name != other.name) return false
        if (format != other.format) return false
        if (modelType != other.modelType) return false
        if (!textureData.contentEquals(other.textureData)) return false
        if (folder != other.folder) return false
        if (isFavorite != other.isFavorite) return false
        if (lastUpdated != other.lastUpdated) return false

        return true
    }

    override fun hashCode(): Int {
        var result = id
        result = 31 * result + name.hashCode()
        result = 31 * result + format.hashCode()
        result = 31 * result + modelType.hashCode()
        result = 31 * result + textureData.contentHashCode()
        result = 31 * result + folder.hashCode()
        result = 31 * result + isFavorite.hashCode()
        result = 31 * result + lastUpdated.hashCode()
        return result
    }
}

File: server/models/SkinProject.js
const { DataTypes } = require('sequelize');
const { sequelize } = require('../config/database');

/**
 * SkinProject — mirrors the Android Room entity exactly:
 *
 *   id          INT  PK autoincrement
 *   name        TEXT
 *   format      TEXT  ("64x32" | "64x64" | "128x128")
 *   modelType   TEXT  ("STEVE" | "ALEX")
 *   textureData BLOB  (raw PNG bytes, base64 over JSON)
 *   folder      TEXT
 *   isFavorite  BOOLEAN
 *   lastUpdated BIGINT (epoch ms, matches System.currentTimeMillis())
 */
const SkinProject = sequelize.define('SkinProject', {
  id: {
    type: DataTypes.INTEGER,
    primaryKey: true,
    autoIncrement: true,
  },
  name: {
    type: DataTypes.STRING,
    allowNull: false,
    validate: { notEmpty: { msg: 'Skin name cannot be empty.' } },
  },
  format: {
    type: DataTypes.STRING,
    allowNull: false,
    defaultValue: '64x64',
    validate: {
      isIn: {
        args: [['64x32', '64x64', '128x128']],
        msg: 'format must be one of: 64x32, 64x64, 128x128',
      },
    },
  },
  modelType: {
    type: DataTypes.STRING,
    allowNull: false,
    defaultValue: 'STEVE',
    validate: {
      isIn: {
        args: [['STEVE', 'ALEX']],
        msg: 'modelType must be STEVE or ALEX',
      },
    },
  },
  // Stored as base64 string in SQLite; Android sends/receives base64
  textureData: {
    type: DataTypes.TEXT('long'),
    allowNull: false,
    comment: 'Base64-encoded PNG texture sheet bytes',
  },
  folder: {
    type: DataTypes.STRING,
    allowNull: false,
    defaultValue: 'My Skins',
  },
  isFavorite: {
    type: DataTypes.BOOLEAN,
    allowNull: false,
    defaultValue: false,
  },
  lastUpdated: {
    type: DataTypes.BIGINT,
    allowNull: false,
    defaultValue: () => Date.now(),
  },
}, {
  tableName: 'skin_projects',
  timestamps: false,
});

module.exports = SkinProject;
